accepted/00000-http-api.md

```diff
@@ -0,0 +1,220 @@
+
+On the HTTP side, Spark allows registration of an endpoint as a "route" per each endpoint method, optionally including a number of named route parameters that construct the route path.
+
+The actual handler methods are designed to use POJOs as parameters. However, Spark expects a specific signature that takes `request` and `response` as parameters for route handlers. A simple wrapper shall unpack the parameters from the request and call the appropriate handler method for each request.
+
+A typical resource-oriented REST API relies on various HTTP verbs to provide contextual information on what type of operation an endpoint is used for. However, the XMLRPC protocol is designed around methods, rather than resources. Therefore, HTTP verbs will be restricted to GET for read-only, and POST for read-write operations regardless of whether they require a payload or not.
+
+Spark provides 3 common ways to attach data to a request. Named route parameters that are part of the base URL, querystring parameters, or the payload itself.
+
+For primitive types, either route parameters or querystring parameters may be used. The desired method shall be specified using a parameter annotation in the handler method signature (explained below).
+
+For complex types, the request payload of type `JSON` must be used. The Spark route wrapper shall unpack the payload using the `GSON` library.
+
+The exposed HTTP routes shall follow the same namespace-method structure with a few extensions:
+
+```
+/rhn/manager/api/namespace/method[/:route/:params][?query=params]
+```
+
+On each request, the Spark route wrapper shall unpack both types of arguments in the request, deserialize the payload, and call the appropriate handler method with unpacked arguments.
+
+To provide a uniform approach to handler registration, Java annotations can be used to specify a method as an endpoint handler. These annotations can be interpreted during initialization to register all the handlers to both frameworks. As a side benefit, all the handler classes in a predetermined package tree can be automatically identified, without the need to add them in a list manually.
+
+Java annotations shall also be used to determine attributes of an endpoint, such as whether it is a read-only method or not.
+
+Below is an example of such an annotation structure:
+
+#### @ApiHandler("namespace") Target=Class
+
+A class marked with this annotation will be registered as the handler of the specified namespace in the XMLRPC framework. The class will further be inspected on initialization to discover any annotated handler methods. Found methods will be registered to Spark as route handlers.
```

Comment:
Unfortunately not. The XMLRPC framework is designed in a way that it recognizes a whole class per endpoint (namespace). The additional level of multiple methods per namespace is our own implementation.

Reply:
Even if the XML-RPC recognizes a whole class per endpoint, couldn't annotations make it easier to match the methods within the namespace?

Reply:
A huge blocker for this is the limitation that we cannot do a simple annotation scan in plain Java. There's a library called `reflections` for this purpose.
But honestly, since we're not adding any new handlers at this point and will use them as-is, implementing this is a bit overkill, so I removed this from the text.